File: 2023/01/01.lua
-- Read input file
local lines = {}
local file = io.open("input.txt", "r")
for line in file:lines() do
    table.insert(lines, line)
end
file:close()

-- Process each line to find calibration values
local calibrationCodes = {}
for _, line in ipairs(lines) do
    local firstDigit, lastDigit
    
    -- Find first digit
    for i = 1, #line do
        local char = line:sub(i,i)
        if tonumber(char) then
            firstDigit = char
            break
        end
    end
    
    -- Find last digit
    for i = #line, 1, -1 do
        local char = line:sub(i,i)
        if tonumber(char) then
            lastDigit = char
            break
        end
    end
    
    -- Combine digits and add to calibration codes
    if firstDigit and lastDigit then
        local calibrationValue = tonumber(firstDigit .. lastDigit)
        table.insert(calibrationCodes, calibrationValue)
    end
end

-- Sum all calibration codes
local sum = 0
for _, code in ipairs(calibrationCodes) do
    sum = sum + code
end

print("Sum of calibration values:", sum)

File: 2023/01/01.js
// Read the input file
const fs = require('fs');
const input = fs.readFileSync('input.txt', 'utf-8');

// Break the file on \n
const lines = input.split('\n');

const wordMap = {
    'one': '1',
    'two': '2',
    'three': '3',
    'four': '4',
    'five': '5',
    'six': '6',
    'seven': '7',
    'eight': '8',
    'nine': '9'
};

const regexMap = new RegExp(Object.keys(wordMap).join('|'), 'g');

// Iterate over each line
const calibrationCodes = lines.map(line => {
    line = line.replace(regexMap, function(match) {
        return wordMap[match];
    });
    console.log(line);

    // Here lineArrays are chacters
    const lineArray = line.split('');

    const firstDigit = lineArray.find(char => !isNaN(char));
    
    // Find the last !isNaN
    const lastDigit = lineArray.reverse().find(char => !isNaN(char));

    if (firstDigit && lastDigit) {
        return parseInt(firstDigit + lastDigit); // '46' -> 46
    }
}).filter(Boolean);

// Sum the calibraion codes
console.log('Calibration codes: ', calibrationCodes);
const sum = calibrationCodes.reduce((acc, curr) => acc + curr, 0);
console.log('Sum: ', sum);